File: src/types/document.ts
/**
 * Document Domain Types & Constants
 * ScholarHub Official Document Repository & Verification Architecture
 */

/**
 * 7 Official Document Categories
 */
export type DocumentCategory =
  | 'Identity'
  | 'Academic'
  | 'Financial'
  | 'Category'
  | 'Banking'
  | 'Personal'
  | 'Application';

export const DOCUMENT_CATEGORIES: DocumentCategory[] = [
  'Identity',
  'Academic',
  'Financial',
  'Category',
  'Banking',
  'Personal',
  'Application',
];

/**
 * 12 Standard Document Types defined by ScholarHub
 */
export type DocumentType =
  // Identity
  | 'aadhaar'
  | 'passport'
  | 'pan'
  // Financial / Category
  | 'income_certificate'
  | 'caste_certificate'
  // Academic
  | 'marksheet'
  | 'bonafide_certificate'
  // Banking
  | 'bank_passbook'
  // Personal
  | 'photograph'
  | 'signature'
  // Application
  | 'sop'
  | 'recommendation_letter';

/**
 * Document Status (Frontend Stage)
 */
export type DocumentStatus = 'UPLOADED' | 'MISSING';

/**
 * Strongly Typed Document Item Model
 */
export interface DocumentItem {
  id: string;
  type: DocumentType;
  name: string;
  category: DocumentCategory;
  status: DocumentStatus;
  fileName?: string;
  fileSize?: string;
  fileSizeBytes?: number;
  uploadedAt?: string;
  expiryDate?: string;
  fileUrl?: string;
  mimeType?: string;
  description?: string;
}

/**
 * Category Metadata for UI rendering
 */
export interface CategoryMetadata {
  category: DocumentCategory;
  icon: string;
  description: string;
}

export const CATEGORY_METADATA: Record<DocumentCategory, CategoryMetadata> = {
  Identity: {
    category: 'Identity',
    icon: '🪪',
    description: 'Government issued identity proofs & citizenships',
  },
  Academic: {
    category: 'Academic',
    icon: '🎓',
    description: 'Grade cards, transcripts & institutional bonafides',
  },
  Financial: {
    category: 'Financial',
    icon: '💵',
    description: 'Income certificates & socio-economic declarations',
  },
  Category: {
    category: 'Category',
    icon: '📜',
    description: 'Affirmative action & quota verification certificates',
  },
  Banking: {
    category: 'Banking',
    icon: '🏦',
    description: 'Bank passbook & cancelled cheques for Direct Benefit Transfer',
  },
  Personal: {
    category: 'Personal',
    icon: '👤',
    description: 'Biometric photograph & signature specimens',
  },
  Application: {
    category: 'Application',
    icon: '📝',
    description: 'Essays, Statements of Purpose & Recommendation Letters',
  },
};

/**
 * File Validation Constants
 */
export const MAX_DOCUMENT_FILE_SIZE_BYTES = 10 * 1024 * 1024; // 10 Megabytes
export const MAX_DOCUMENT_FILE_SIZE_MB = 10;

export const SUPPORTED_FILE_EXTENSIONS = ['.pdf', '.jpg', '.jpeg', '.png'] as const;

export const SUPPORTED_MIME_TYPES = [
  'application/pdf',
  'image/jpeg',
  'image/jpg',
  'image/png',
] as const;

File: backend/app/schemas/document.py
import uuid
from datetime import date, datetime
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class DocumentTypeEnum(str, Enum):
    """Controlled document types matching the ScholarHub frontend specifications."""
    AADHAAR = "aadhaar"
    PASSPORT = "passport"
    PAN = "pan"
    INCOME_CERTIFICATE = "income_certificate"
    CASTE_CERTIFICATE = "caste_certificate"
    MARKSHEET = "marksheet"
    BONAFIDE_CERTIFICATE = "bonafide_certificate"
    BANK_PASSBOOK = "bank_passbook"
    PHOTOGRAPH = "photograph"
    SIGNATURE = "signature"
    SOP = "sop"
    RECOMMENDATION_LETTER = "recommendation_letter"
    OTHER = "other"


class DocumentCategoryEnum(str, Enum):
    """Document categories matching the ScholarHub frontend specifications."""
    IDENTITY = "Identity"
    ACADEMIC = "Academic"
    FINANCIAL = "Financial"
    CATEGORY = "Category"
    BANKING = "Banking"
    PERSONAL = "Personal"
    APPLICATION = "Application"
    OTHER = "Other"


class DocumentResponse(BaseModel):
    """Metadata response schema for a student document."""
    id: uuid.UUID
    user_id: uuid.UUID
    type: str
    category: str
    file_name: Optional[str] = None
    file_size: Optional[int] = None
    file_url: Optional[str] = None
    uploaded_at: Optional[datetime] = None
    expiry_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DocumentListResponse(BaseModel):
    """Paginated list of student documents."""
    items: List[DocumentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DocumentDeleteResponse(BaseModel):
    """Response confirming document deletion."""
    message: str = "Document deleted successfully"
    document_id: uuid.UUID
